go/vt/mysqlctl/binlogs_gtid.go

```diff
@@ -71,43 +71,44 @@ func ChooseBinlogsForIncrementalBackup(
 		}
 		if prevGTIDsUnion == nil {
-			prevGTIDsUnion = prevPos.GTIDSet
+			prevGTIDsUnion = previousGTIDsPos.GTIDSet
 		} else {
-			prevGTIDsUnion = prevGTIDsUnion.Union(prevPos.GTIDSet)
+			prevGTIDsUnion = prevGTIDsUnion.Union(previousGTIDsPos.GTIDSet)
 		}
 
-		containedInFromPos := lookFromGTIDSet.Contains(prevPos.GTIDSet)
-		// The binary logs are read in-order. They are build one on top of the other: we know
-		// the PreviousGTIDs of once binary log fully cover the previous binary log's.
-		if containedInFromPos {
-			// All previous binary logs are fully contained by backupPos. Carry on
-			continue
-		}
-		// We look for the first binary log whose "PreviousGTIDs" isn't already fully covered
-		// by "backupPos" (the position from which we want to create the inreemental backup).
+		// The binary logs are read in-order. They expand. For example, we know
+		// Previous-GTIDs of binlog file 0000018 contain those of binlog file 0000017.
+		// We look for the first binary log whose Previous-GTIDs isn't already fully covered
+		// by "backupPos" (the position from which we want to create the incremental backup).
 		// That means the *previous* binary log is the first binary log to introduce GTID events on top
 		// of "backupPos"
-		if i == 0 {
-			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "the very first binlog file %v has PreviousGTIDs %s that exceed given incremental backup pos. There are GTID entries that are missing and this backup cannot run", binlog, prevPos)
+		if backupFromGTIDSet.Contains(previousGTIDsPos.GTIDSet) {
+			// Previous-GTIDs is contained by backupPos. So definitely all binlogs _prior_ to
+			// this binlog are not necessary. We still don't know about _this_ binlog. We can't tell yet if
+			// _this_ binlog introduces new GTID entries not covered by the last backup pos. But we will only
+			// know this when we look into the _next_ binlog file's Previous-GTIDs.
+			continue
 		}
-		if unionPreviousGTIDs {
-			prevPos.GTIDSet = prevGTIDsUnion
+		if i == 0 {
+			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "Required entries have been purged. Oldest binary log %v expects entries not found in back up pos. Expected pos=%v", binlog, previousGTIDsPos)
 		}
-		if !prevPos.GTIDSet.Contains(lookFromGTIDSet) {
-			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "binary log %v with previous GTIDS %s neither contains requested GTID %s nor contains it. Backup cannot take place", binlog, prevPos.GTIDSet, lookFromGTIDSet)
+		if !prevGTIDsUnion.Union(purgedGTIDSet).Contains(backupFromGTIDSet) {
+			return nil, "", "", vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION,
+				"Mismatching GTID entries. Requested backup pos has entries not found in the binary logs, and binary logs have entries not found in the requested backup pos. Requested pos=%v, binlog pos=%v",
```

Comment:
Is it `and binary logs` or `or binary logs`?
In other words, does the check imply both conditions have occurred, or that at least one of them has occurred?
You could replace `and` with `and/or`.

Reply:
It's an `and`, as described. It's like a Venn diagram of two circles which overlap, but neither fully contains the other.

Reply:
Added `Neither fully contains the other` to the message. I hope it clarifies.